### PR TITLE
Add best_score to early stopping callback

### PR DESCRIPTION
**Pull request by mratsim**
PR to close https://github.com/Microsoft/LightGBM/issues/329

### CODE REVIEW

File: python-package/lightgbm/engine.py [hunk 1]
```diff
@@ -196,8 +196,10 @@ def train(params, train_set, num_boost_round=100,
             break
     if booster.attr('best_iteration') is not None:
         booster.best_iteration = int(booster.attr('best_iteration')) + 1
+        booster.best_score = float(booster.attr('best_score'))
     else:
         booster.best_iteration = -1
+        booster.best_score = -1
```
Comment (wxchan) [inline python-package/lightgbm/engine.py]: don't use -1 here, -1 can be 'real' best_score; `float('inf')` or `None` might be better

**Comment by wxchan** [comment]
it'll be better to merge previous commits into last one with `git rebase`.

File: python-package/lightgbm/callback.py [hunk 2]
```diff
@@ -14,10 +14,13 @@ class EarlyStopException(Exception):
     ----------
     best_iteration : int
         The best iteration stopped.
+    best_score : float
```
Comment (wxchan) [inline python-package/lightgbm/callback.py]: there might be multiple eval datasets and multiple eval metrics, you cannot use only one floating number for best_score

File: python-package/lightgbm/callback.py [hunk 3]
```diff
@@ -164,6 +167,7 @@ def early_stopping(stopping_rounds, verbose=True):
     best_iter = []
     best_msg = []
     cmp_op = []
+    score_list = []
```
Comment (wxchan) [inline python-package/lightgbm/callback.py]: I think you should use dict here, same format as record_evaluation: {dataset name: {eval metric name: score}}, otherwise user can't know the order of those metrics

**Comment by wxchan** [comment]
can you also add a test case in tests/python_package_test/test_engine.py or test_sklearn.py? one is ok. you can use template function or not as you like. also update the corresponding documents

**Comment by wxchan** [comment]
check #369 

I made some changes to early stopping, you don't need to use set_attr to pass best_score; and I add a test for early stopping, you can add best_score to it

btw, score_list and best_msg actually store the same information, perhaps you can merge them into one, but it's trivial, you can leave it as it is

**Comment by guolinke** [comment]
@wxchan is this PR ready?

**Comment by wxchan** [comment]
it's not finished

**Comment by Laurae2** [comment]
@guolinke @wxchan 

I talked with @mratsim and he will rebase and add the test today/tomorrow

**Comment by mratsim** [comment]
Hello @Laurae2 @wxchan @guolinke. Unfortunately I'm quite taken at the moment.
I'm closing this PR for now.

Is that fine with you?

**Comment by wxchan** [comment]
ok, I might update this sometime.
